trl/trainer/grpo_trainer.py

```diff
@@ -388,21 +411,49 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
         if self.args.use_vllm:
             # First, have main process load weights if needed
             if self.state.global_step != self._last_loaded_step:
-                with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
-                    if isinstance(unwrapped_model, PeftModel):
-                        unwrapped_model = copy.deepcopy(unwrapped_model)
-                        unwrapped_model.merge_and_unload()
-                        state_dict = unwrapped_model.state_dict()
-
-                if self.accelerator.is_main_process:
-                    llm_model = self.llm.llm_engine.model_executor.driver_worker.model_runner.model
-                    llm_model.load_weights(state_dict.items())
+                if self.use_lora:
+                    # vLLM supports natively loading LoRA weights dynamically at runtime, so 
+                    # save weights to a temporary directory and prepare a request for the vLLM runtime.
+                    intermediate_checkpoint_dir = os.path.join(
+                        self.args.output_dir, f".vllm-lora-checkpoint-{self.state.global_step}"
+                    )
+                    model.save_pretrained(intermediate_checkpoint_dir)
+                    self.tokenizer.save_pretrained(intermediate_checkpoint_dir)
+                    self.lora_request = LoRARequest(
+                        f"vllm-lora-checkpoint-{self.state.global_step}",
+                        self.state.global_step,
+                        lora_path=intermediate_checkpoint_dir,
+                    )
+
+                    # Cleanup previous checkpoint
+                    last_global_step = self.state.global_step - 1
+                    last_checkpoint_dir = os.path.join(
+                        self.args.output_dir, f".vllm-lora-checkpoint-{last_global_step}"
+                    )
+                    if os.path.exists(last_checkpoint_dir):
+                        shutil.rmtree(last_checkpoint_dir)
+                else:
+                    # If not using LoRA, we load the full model weights in vLLM
+                    with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
+                        if isinstance(unwrapped_model, PeftModel):
+                            unwrapped_model = copy.deepcopy(unwrapped_model)
+                            unwrapped_model.merge_and_unload()
```

Comment:
Similar comment as for the other PR: Consider using `model.merge_adapter`, cleaning up the state dict, then `model.unmerge_adapter`, instead of creating a whole copy of the model.